File: LinkedLists/Java/ReverseLL.java
package LinkedLists.Java;

public class ReverseLL {
    public static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    private static Node convertArrayToLinkedList(int[] arr) {
        if (arr.length == 0) return null;
        Node head = new Node(arr[0]);
        Node current = head;
        for (int i = 1; i < arr.length; i++) {
            current.next = new Node(arr[i]);
            current = current.next;
        }
        return head;
    }

    public static Node reverseList(Node head) {
        // Time Complexity: O(N) where N is the number of nodes in the linked list
        // Space Complexity: O(1) since we are not using any extra space except for pointers
        Node prev = null;
        Node temp = head;

        while (temp != null) {
            Node nextNode = temp.next; // Store the next node
            temp.next = prev; // Reverse the link
            prev = temp; // Move prev to current
            temp = nextNode; // Move to the next node
        }
        return prev; // New head of the reversed list
    }

    private static void printLinkedList(Node head) {
        while (head != null) {
            System.out.print(head.data + "->");
            head = head.next;
        }
        System.out.println("null");
    }

    private static Node reverseListBruteForce(Node head) {
        // Time Complexity: O(N) where N is the number of nodes in the linked list
        // Space Complexity: O(N) since we are using a stack to store the nodes
        if (head == null || head.next == null) return head;

        java.util.Stack<Node> stack = new java.util.Stack<>();
        Node current = head;

        // Push all nodes onto the stack
        while (current != null) {
            stack.push(current);
            current = current.next;
        }

        // Pop nodes from the stack to reverse the linked list
        Node newHead = stack.pop();
        current = newHead;

        while (!stack.isEmpty()) {
            current.next = stack.pop();
            current = current.next;
        }
        current.next = null; // Set the last node's next to null

        return newHead; // Return the new head of the reversed list
    }

    private static Node reverseListRecursive(Node head) {
        // Time Complexity: O(N) where N is the number of nodes in the linked list
        // Space Complexity: O(N) due to recursion stack space
        if (head == null || head.next == null) return head;

        // Solve the list for the smaller problem
        // and then reverse the link
        // between the current node and the next node
        Node newHead = reverseListRecursive(head.next);
        Node nextNode = head.next; // Store the next node
        nextNode.next = head; // Reverse the link
        head.next = null; // Set the next of the current node to null

        return newHead; // Return the new head of the reversed list
    }

    public static void main(String[] args) {
        Node head = convertArrayToLinkedList(new int[]{1, 2, 3, 4, 5});
        System.out.println("Original Linked List:");
        printLinkedList(head);

        Node reversedHead = reverseList(head);
        System.out.println("Reversed Linked List:");
        printLinkedList(reversedHead);
        
        Node bruteForceReversedHead = reverseListBruteForce(reversedHead);
        System.out.println("Reversed Linked List using Brute Force:");
        printLinkedList(bruteForceReversedHead);
        Node recursiveReversedHead = reverseListRecursive(bruteForceReversedHead);
        System.out.println("Reversed Linked List using Recursion:");
        printLinkedList(recursiveReversedHead);
        // Note: The original head is now reversed, so we can reverse it again to get back the original order
    }
}

File: LinkedLists/Python/reverseLL.py
class Node:
    def __init__(self,data1,next1 = None):
        self.data = data1
        self.next = next1
        
def convertarr2LL(arr):
    head = Node(arr[0])
    mover = head
    
    for i in range(1,len(arr)):
        tmp = Node(arr[i])
        mover.next = tmp
        mover = tmp
        
    return head
    
def reverseLL(head):
    if not head or not head.next:
        return head
    tmp = head
    dummy = None
    while tmp:
        front = tmp.next
        tmp.next = dummy
        dummy = tmp
        tmp = front
    
    return dummy
    
def reverseLL2(head):
    if not head or not head.next:
        return head
        
    newHead = reverseLL2(head.next)
    front = head.next
    front.next = head
    head.next = None
    return newHead
        
def printLL(head):
    while head:
        print(head.data,end=' ')
        head = head.next
    print('None')
    
def main():
    arr = [4,8,6,9]
    res = convertarr2LL(arr)
    rev = reverseLL(res)
    print("The iterative linked list is: ",end='')
    printLL(rev)
    rev2 = reverseLL2(rev)
    print("The recursive linked list is: ",end='')
    printLL(rev2)
    
if __name__ == "__main__":
    main()